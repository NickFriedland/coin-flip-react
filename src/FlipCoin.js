import React, { Component } from 'react';
import Coin from './Coin';
import Heads from './coin-heads.png';
import Tails from './coin-tails.png';

class FlipCoin extends Component {
  constructor(props) {
    super(props);
    this.state = {
      numHeads: 0,
      numTails: 0,
      coinState: 0
    };
    this.flipCoin = this.flipCoin.bind(this);
  }
  // the two imgs addresses
  static defaultProps = {
    urls: [Heads, Tails]
  };

  // callback function for flip

  flipCoin() {
    this.setState(
      {
        coinState: Math.floor(Math.random() * 2)
      },
      this.countOutcome
    );
  }

  countOutcome() {
    this.setState(curState => {
      return curState.coinState
        ? { numTails: curState.numTails + 1 }
        : { numHeads: curState.numHeads + 1 };
    });
  }
  // random select 0 or 1 for coin
  // count head or tail for the

  render() {
    return (
      <div>
        <Coin url={this.props.urls[this.state.coinState]} />
        <button onClick={this.flipCoin} />
        <p>
          Out of {this.state.numHeads + this.state.numTails} flips, there have
          been {this.state.numHeads} heads and {this.state.numTails} tails
        </p>
      </div>
    );
  }
}

export default FlipCoin;
