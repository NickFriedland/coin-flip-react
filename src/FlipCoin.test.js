import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import FlipCoin from './FlipCoin';

// smoke test
describe('<FlipCoin />', function() {
  it('renders without crashing', function() {
    shallow(<FlipCoin />);
  });

  //snapshot test
  it('matches snapshot', function() {
    const wrapper = shallow(<FlipCoin />);
    const serialized = toJson(wrapper);
    expect(serialized).toMatchSnapshot();
  });
  /* enzyme selector tests */

  // default state

  it('has a default state with numHeads: 0, numTails: 0, coinState: 0', function() {
    const wrapper = shallow(<FlipCoin />);
    // has default state of count: 0
    expect(wrapper.state().numHeads).toBe(0);
    expect(wrapper.state().numTails).toBe(0);
    expect(wrapper.state().coinState).toBe(0);
  });

  // test flip, find the id and simulate to click the button

  it('flip the coin', function() {
    /* the following is the same as above */
    const wrapper = shallow(<FlipCoin />);
    expect(wrapper.state().numHeads).toBe(0);
    expect(wrapper.state().numTails).toBe(0);
    expect(wrapper.state().coinState).toBe(0);

    const flip = wrapper.find('#flip');
    flip.simulate('click');

    expect(wrapper.state().numHeads + wrapper.state().numTails).toBe(1);
  });
});
