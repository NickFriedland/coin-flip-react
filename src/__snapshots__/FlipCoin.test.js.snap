// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlipCoin /> matches snapshot 1`] = `
<div>
  <Coin
    url="coin-heads.png"
  />
  <button
    id="flip"
    onClick={[Function]}
  />
  <p>
    Out of 
    0
     flips, there have been 
    0
     heads and 
    0
     tails
  </p>
</div>
`;
